scriptreplay: add interactive playback controls

jNullj opened the pull request:
Adds interactive playback controls based of #2999 .

Status:
- [x] Pausing and resuming playback: space to toggle.
- [x] Adjusting playback speed: up/down (arrows), +/- 0.1x steps
- [x] Stepwise navigation: right (arrow) for step forward
- [x] Add this info and key-bindings to docs/man.

I did not add all suggested in #2999 , These are what i think the most important.
Adding a backward step is non trivial without rewriting lots of how the playback works, and most terminals allow to scroll back the stream anyway.
Speed for holding the key-bind is also self emerging from the `inputDelay`.

So i think its best to attempt to merge this basic form so it ships the core features with less complex changes.

jNullj commented:
I intend on changing that, pause currently ignores the current step.
I will change the loop to check for input every `inputDelay` to avoid blocking from current step.
This will also provide a more responsive control for all controls.

jNullj commented on term-utils/scriptreplay.c:
```diff
@@ -308,9 +320,69 @@
+								break;
+						}
+						break;
+				}
```
Refactored this with https://github.com/util-linux/util-linux/pull/3015/commits/6cc9d27d9851090d118c0851dcf0aeaf869da132

karelzak commented on term-utils/scriptreplay.c:
```diff
@@ -143,9 +152,12 @@ int
 {
 	static const struct timeval mindelay = { .tv_sec = 0, .tv_usec = 100 };
+	static const struct timeval inputDelay = { .tv_sec = 0, .tv_usec = 100000 };
+	struct timeval stepDelay = { 0, 0 };
```
Please keep the variable names consistent. We rarely use C++-like names like "fooBar;" it seems better to use "inputdelat" or "input_delay."

jNullj replied:
Fixed with https://github.com/util-linux/util-linux/pull/3015/commits/28d7c58404ec8b38021f90bd8930783f0279ebe0

karelzak commented on term-utils/scriptreplay.c:
```diff
@@ -308,9 +320,64 @@ main(int argc, char *argv[])
+	isterm = setterm(&saved, &saved_flag);
 
 	do {
+		switch (fgetwc(stdin)) {
```
There is include/fgetwc_or_err.h with fgetwc_or_err(). I guess it can increase the reliability of the code.

jNullj replied:
IMO we should not use fgetwc_or_err as we expect fgetwc to have many instances of WEOF with errno !=0 when no user input is added to the stream.
```c
// section from fgetwc_or_err
if (ret == WEOF && errno != 0)
		err(EXIT_FAILURE, _("fgetwc() failed"));
```
We run at non-blocking mode and this results in this error showing up every time no input is found `getwc() failed: Resource temporarily unavailable`, which is very common and repeats often during a replay.

karelzak replied:
Good point, I forgot about the non-blocking mode ...

karelzak replied:
BTW, why do we need widechar (fgetwc() and wchat_t)  here? I guess it will work with arbitrary function like fgetc().

karelzak commented on term-utils/scriptreplay.c:
```diff
@@ -308,9 +320,64 @@ main(int argc, char *argv[])
+			break;
+
+		if (replay_get_is_paused(setup))
+		{
```
Please follow the current coding style. It means "if ( ) {" rather than "if ( )\n{".
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces

We also do not use "//" for comments :-)

jNullj replied:
Fixed with https://github.com/util-linux/util-linux/pull/3015/commits/0bfd75dbf413be85049e0c89551f208cfa64a2ea

karelzak commented:
https://github.com/util-linux/util-linux/actions/runs/9228958044/job/25445611203?pr=3015
```
term-utils/scriptreplay.c: In function ‘main’:
term-utils/scriptreplay.c:330:5: error: a label can only be part of a statement and a declaration is not a statement
  330 |     int first_char = fgetc(stdin);
      |     ^~~
make[2]: *** [Makefile:11104: term-utils/scriptreplay.o] Error 1
```

jNullj commented:
My bad, used gcc at my setup.
I noticed we already have a generic char to use in another loop in main `int ch` so i used it to save up on memory a bit. I think it fits better and also makes it compatible with clang.
Fixed with https://github.com/util-linux/util-linux/pull/3015/commits/c2c16fdb72a83aef157813045738ed5c96513309 and tested both with gcc and clang this time (i have clang-17 tho).

Also noticed the switch-case coding style is not as recommended by the linux kernel and fixed the indent with https://github.com/util-linux/util-linux/pull/3015/commits/d616c2fb43e5d302fbc6af2d335c0574c63af9c6

karelzak commented on term-utils/scriptreplay.c:
```diff
@@ -122,10 +125,15 @@ appendchr(char *buf, size_t bufsz, int c)
+	*saved_flag = fcntl(STDIN_FILENO, F_GETFL);
+	if (*saved_flag == -1)
+		err(EXIT_FAILURE, _("unexpected fcntl failure"));
+	fcntl(STDIN_FILENO, F_SETFL, O_NONBLOCK);
```
@jNullj  -- Should this be implemented as:
```
     fcntl(STDIN_FILENO, F_SETFL, saved_flag | O_NONBLOCK);
```
It means adding O_NONBLOCK to exiting flags rather than resetting all.

jNullj replied:
Good catch! The current code does rewrite the flags in that int insted of adding the nonblock flag.
My only addition would be using `*saved_flag` here as `saved_flag` is *int and not int.

```suggestion
	fcntl(STDIN_FILENO, F_SETFL, *saved_flag | O_NONBLOCK);
```

Should i make a new commit to this branch and open a PR to add this fix?

karelzak replied:
Fixed by https://github.com/util-linux/util-linux/commit/41e7686c9ad1ea7892b9d8941c266869bf6a28dd
